File: C++/CreationalDesignPattern/ObserverDesign_pattern.cpp
#include <algorithm>
#include <iostream>
#include <memory>
#include <string>
#include <vector>

struct Follower {
    virtual void notify(const std::string& message) = 0;
    virtual ~Follower() = default;
};

struct Subject {
    virtual void addObserver(std::shared_ptr<Follower> observer) = 0;
    virtual void removeObserver(std::shared_ptr<Follower> observer) = 0;
    virtual ~Subject() = default;
};

class WhatsAppBroadCast : public Subject {
    std::vector<std::shared_ptr<Follower>> followers;

public:
    void addObserver(std::shared_ptr<Follower> observer) override {
        followers.push_back(observer);
    }

    void removeObserver(std::shared_ptr<Follower> observer) override {
        followers.erase(std::remove(followers.begin(), followers.end(), observer), followers.end());
    }

    void addFollower(std::shared_ptr<Follower> follower) { addObserver(follower); }

    void removeFollower(std::shared_ptr<Follower> follower) { removeObserver(follower); }

    void sendMessage(const std::string& message) {
        for (auto& follower : followers) {
            follower->notify(message);
        }
    }
};

class Prateek : public Follower {
public:
    void notify(const std::string& message) override {
        std::cout << " Prateek Received Message: " << message << std::endl;
    }
};

class Abhinav : public Follower {
public:
    void notify(const std::string& message) override {
        std::cout << "Abhinav Received Message: " << message << std::endl;
    }
};

int main() {
    WhatsAppBroadCast whatsAppBroadCast;
    whatsAppBroadCast.addFollower(std::make_shared<Prateek>());
    whatsAppBroadCast.addFollower(std::make_shared<Abhinav>());

    whatsAppBroadCast.sendMessage("pradeep is teaching Observer pattern");

    return 0;
}

File: Python/CreationalDesignPattern/ObserverDesign_pattern.py
from abc import ABC, abstractmethod
from typing import List


class Subject(ABC):
    @abstractmethod
    def add_observer(self, observer):
        pass

    @abstractmethod
    def remove_observer(self, observer):
        pass


class Follower(ABC):
    @abstractmethod
    def notify(self, message: str):
        pass


class WhatsAppBroadCast:
    def __init__(self):
        self.followers: List[Follower] = []

    def add_follower(self, follower: Follower):
        self.followers.append(follower)

    def remove_follower(self, follower: Follower):
        self.followers.remove(follower)

    def send_message(self, message: str):
        for follower in self.followers:
            follower.notify(message)


class Prateek(Follower):
    def notify(self, message: str):
        print(f" Prateek Received Message: {message}")


class Abhinav(Follower):
    def notify(self, message: str):
        print(f"Abhinav Received Message: {message}")


def main():
    whats_app_broad_cast = WhatsAppBroadCast()
    whats_app_broad_cast.add_follower(Prateek())
    whats_app_broad_cast.add_follower(Abhinav())

    whats_app_broad_cast.send_message("pradeep is teaching Observer pattern")


if __name__ == "__main__":
    main()
